File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/format/DurationHms.java
package com.freight.format;

/** Formatter: duration hms. */
public final class DurationHms implements Formatter {

    private static final String BASE36_DIGITS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    @Override
    public String name() {
        return "duration_hms";
    }

    @Override
    public String apply(long value) {
        boolean negative = value < 0L;
        long absolute = negative ? -value : value;
        return (negative ? "-" : "") + String.format(java.util.Locale.ROOT, "%02d:%02d:%02d",
                absolute / 3600L, (absolute / 60L) % 60L, absolute % 60L);
    }
}

File: freight-triage-polyglot-ledger-recovery/environment/native/src/format/duration_hms.cpp
#include "freight/formats.h"

#include <algorithm>
#include <cstdio>

namespace freight {

// Formatter: duration hms.
std::string fmt_duration_hms(long long value) {
  bool negative = value < 0;
  long long absolute = negative ? -value : value;
  char buffer[64];
  std::snprintf(buffer, sizeof(buffer), "%s%02lld:%02lld:%02lld", negative ? "-" : "",
                absolute / 3600, (absolute / 60) % 60, absolute % 60);
  return std::string(buffer);
}

}  // namespace freight
